File: parserprototype/parser.py
import sys

class Parser():
	def infix_to_postfix(self, expr):
	    stack = list()
	    postfix_expr = list()
	    for c in expr:
	 	if c > '0' and c < '9':
			#Add terminal to the postfix expression
			postfix_expr.append(c)
		if c=='*' or c=='/':
			#Higher precedence operators are pushed onto the stack
			stack.append(c)
		if c=='+' or c=='-':
			#Pop out operators with higher precedence from the stack and append it to the postfix expression, then push lower order operators on the stack
			if len(stack) == 0:
				stack.append(c)
				continue
			else:
			        s_top=stack.pop()
				if(s_top=='*' or s_top=='/'):
					postfix_expr.append(s_top)
				else:
					stack.append(s_top)
				while( len(stack) >= 0 ):
					s_top=stack.pop()
					if(s_top=='*' or s_top=='/'):
						postfix_expr.append(s_top)
					else:
						stack.append(s_top)
						break
				stack.append(c)
	    while(len(stack) is not 0):
		postfix_expr.append(stack.pop())
	    return postfix_expr
	

	def postfix_parse(self, expr):
		stack = list()
		for c in expr:
	            if c >'0' and c <'9':
                        #Add terminal to stack
		        stack.append(int(c))
		    elif c=='+':
	                #Pop top two values on the stack, evaluate the subexpression and push the result back onto the stack
		  	stack.append(stack.pop()+stack.pop())
		    elif c=='*':
			stack.append(stack.pop()*stack.pop())
		    elif c=="/":
			den = stack.pop()
			stack.append(stack.pop()/den)
		    elif c=='-':
			stack.append(stack.pop() - stack.pop())

		#If the number of terminals exceed the operators, throw an error
		if len(stack) != 1:
			print 'Check expression'
			return 1
		#Print the evaluated expression
		return stack[0]

print 'Enter expression in infix format'
inp_expr = sys.stdin.readline()
parser = Parser()
postfix_expr = parser.infix_to_postfix(inp_expr)
print parser.postfix_parse(postfix_expr)

File: parserprototype/src/parser.rs

pub fn infix_to_postfix(expr: String) -> Vec<String> {

    // Maintain a stack of operators and postfix string of operands
    let mut stack: Vec<char> = Vec::new();
    let mut postfix_expr: Vec<String> = Vec::new();
    let mut operand_stack: VecDeque<char> = VecDeque::new();

    for c in expr.chars() {

	if let Tok = c {
	    // Append terminal to postfix string
	    operand_stack.push_back(c);
        }

	else if c=='*' {
            let mut operand_str = "".to_string();
            while !operand_stack.is_empty() {
                operand_str.push(operand_stack.pop_front().unwrap())
            }
            postfix_expr.push(operand_str);

	    // Pop stack until an operator of lower precedence is found, append the popped operators to the postfix string
	    // This is done to give priority for division over multiplication
            if stack.len() == 0 {
                stack.push(c);
                continue;
            }
            else {
                let mut s_top = stack.pop().unwrap();
                if s_top=='*' || s_top=='/' {
                    postfix_expr.push(s_top.to_string());
                }
                else {
                    stack.push(s_top);
                }
                while stack.len() > 0 {
                    s_top = stack.pop().unwrap();
                    if s_top=='*' || s_top=='/' {
                        postfix_expr.push(s_top.to_string());
                    }
                    else {
                        stack.push(s_top);
                        break;
                    }
                }
            }
	    stack.push(c);
	}

	else if c=='/' {
            let mut operand_str = "".to_string();
            while !operand_stack.is_empty() {
                operand_str.push(operand_stack.pop_front().unwrap())
            }
            postfix_expr.push(operand_str);
	    stack.push(c);
        }

	else if c=='+' || c=='-' {

	    let mut operand_str = "".to_string();
            while !operand_stack.is_empty() {
                operand_str.push(operand_stack.pop_front().unwrap())
            }
            postfix_expr.push(operand_str);

            // Pop stack until an operator of equal precedence is found, append the popped operators to the postfix string
            // This is done to give priority for multiplication and division
	    if stack.len() == 0 {
	        stack.push(c);
		continue;
	    }
	    else {
		let mut s_top = stack.pop().unwrap();
		if s_top=='*' || s_top=='/' {
		    postfix_expr.push(s_top.to_string());
		}
		else {
		    stack.push(s_top);
		}
		while stack.len() > 0 {
		    s_top = stack.pop().unwrap();
		    if s_top=='*' || s_top=='/' {
			postfix_expr.push(s_top.to_string());
		    }
		    else {
			stack.push(s_top);
		        break;
		    }
		}
		stack.push(c);
	    }
	}
    }
    let mut operand_str = "".to_string();
    while !operand_stack.is_empty() {
        operand_str.push(operand_stack.pop_front().unwrap())
    }
    postfix_expr.push(operand_str);

    //Append all the remaining operators to the postfix string
    while stack.len() > 0 {
	postfix_expr.push((stack.pop().unwrap().to_string()));
    }
    //return postfix_expr.iter().cloned().collect::<String>();
    return postfix_expr;
}

pub fn postfix_parse(expr: Vec<String>) -> i32{
    //Maintain a stack of terminals
    let mut stack: Vec<i32> = Vec::new();

    for c in expr.iter() {
        //Push terminals onto stack after type casting them to i32
	if c.chars().nth(0).unwrap() >='0' && c.chars().nth(0).unwrap() <='9' {
	    stack.push(i32::from_str(c).unwrap());
        }

	else if c=="+" || c=="*" || c=="-" || c=="/" {
	    let e1 = stack.pop().unwrap();
	    let e2 = stack.pop().unwrap();
	    if c=="+" {
 	        stack.push(e1 + e2);
            }
	    else if c=="*" {
	        stack.push(e1 * e2);
	    }
	    else if c=="/" {
	        stack.push(e2/e1);
	    }
	    else if c=="-" {
	        stack.push(e2 - e1);
	    }
        }
    }
    if stack.len() > 1 {
	println!("Check expression");
	return 1
    }
    return stack.pop().unwrap();
}
